arrow-libs/fx/arrow-fx-coroutines/src/commonMain/kotlin/arrow/fx/coroutines/Resource.kt

```diff
@@ -354,24 +351,25 @@ public fun <A> resource(block: suspend ResourceScope.() -> A): Resource<A> = blo
  * <!--- KNIT example-resource-06.kt -->
  */
 @ScopeDSL
-public suspend fun <A> resourceScope(action: suspend ResourceScope.() -> A): A {
+@Suppress("REDUNDANT_INLINE_SUSPEND_FUNCTION_TYPE")
+public suspend inline fun <A> resourceScope(action: suspend ResourceScope.() -> A): A {
   val scope = ResourceScopeImpl()
-  val a: A = try {
+  var finished = false
+  return try {
     action(scope)
   } catch (e: Throwable) {
-    val ex = if (e is CancellationException) ExitCase.Cancelled(e) else ExitCase.Failure(e)
-    val ee = withContext(NonCancellable) {
-      scope.cancelAll(ex, e) ?: e
+    finished = true
+    scope.cancelAll(ExitCase.ExitCase(e))
+    throw e
+  } finally {
+    if (!finished) {
+      scope.cancelAll(ExitCase.Completed)
     }
-    throw ee
   }
-  withContext(NonCancellable) {
-    scope.cancelAll(ExitCase.Completed)?.let { throw it }
```

Comment:
We should add a test for this because I'm pretty sure that suppressed exceptions would've been stripped here as well, but it should work properly now

Reply:
Added!